### PR TITLE
Increase severity of JSDoc to error.

### LINKED ISSUE #18458: eslint-plugin: Increase severity of JSDoc rules to error
Previously: https://github.com/WordPress/gutenberg/pull/16869#discussion_r341168788

We should generally seek to avoid any [`'warn'` severity](https://eslint.org/docs/user-guide/configuring#configuring-rules) in our rulesets. If we have a rule, it should be enforced strictly, else it shouldn't be a rule. Exceptions are to be made using [inline comments](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments).

**Task:** In [the `jsdoc` ruleset](https://github.com/WordPress/gutenberg/blob/master/packages/eslint-plugin/configs/jsdoc.js), all custom rule configurations should be changed from `'warn'` to `'error'`. Similarly, if we are extending `'plugin:jsdoc/recommended'`, we should override all rules with equivalents where the severity is changed from `'warn'` to `'error'`.

Reference for `'jsdoc/recommended'` rules: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/src/index.js#L34

### PR DESCRIPTION
**Pull request by sainthkh**
* Closes #18458

## Description

* Updated `eslint-plugin-jsdoc` from 15.8.0 to 21.0.0.
* Increase severity of JSDoc to error.

## How has this been tested?

* `npm run lint-js` command will not show any `warning`s.
* Env: Ubuntu 18.04 

## Notes about decisions

AS it is written in #18458, eslint-plugin-jsdoc cannot parse some TypeScript types like `[key:string]: any`. So, I fixed the issues like these:

* `[key:string]: any`: Mostly, `Record<string, any>`
* Added descriptions
* If all of above aren't possible, I used `/* eslint-disable */`, `/* eslint-enable */` pair. Unfortunately, as JSDoc is written in comment, `// eslint-disable-line` doesn't work. 

## Screenshots 

N/A

## Types of changes

Minor JSDoc comment changes. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] My code has proper inline documentation. 
- [N/A] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
@dsifford and @aduth – what do you think about this proposal?

**Comment by dsifford** [comment]
I'll leave the decision to @aduth as he has more skin in the game than I do here.

**Comment by aduth** [comment]
I've not yet had time to review this in detail, but I'm of course happy to see these warnings bumped to errors.

What I'm not sure about on a cursory glance is whether we want to be changing the existing types to `Record`, since this is a lossy transition ([see example](https://github.com/WordPress/gutenberg/pull/20427/files#diff-106913d5079b57c6032c0775292d8d9dL90-R90), we lose the `'settings'` name). In my mind, I had considered that we simply add inline ESLint configuration to temporarily disable the rules-checking, until such time that the upstream dependencies are updated to support the syntax (per https://github.com/WordPress/gutenberg/issues/18458#issuecomment-576749575, tracked at https://github.com/jsdoctypeparser/jsdoctypeparser/issues/50).

**Comment by sainthkh** [comment]
Rebased.

**Comment by aduth** [comment]
@sainthkh What do you think about my remarks at https://github.com/WordPress/gutenberg/pull/20427#issuecomment-591624655 ? I'm not entirely comfortable with the changes to use `Record`.

File: packages/block-serialization-default-parser/src/index.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ let offset;
 let output;
 let stack;
 
+/* eslint-disable jsdoc/check-values */
```
Comment (aduth) [inline packages/block-serialization-default-parser/src/index.js]: What was the warning that was happening when this wasn't here? Is it not fixed by the changes to the `@type` line?

File: packages/components/src/text/mixins.js [hunk 2]
```diff
@@ -137,7 +137,7 @@ const variant = ( variantName ) => {
 
 /**
  * @typedef {Object} TextProps
- * @property {TextVariant} variant
+ * @property {TextVariant} variant - one of TextVariant to be used
```
Comment (aduth) [inline packages/components/src/text/mixins.js]: WordPress documentation standards don't use the `-`

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/

```suggestion
 * @property {TextVariant} variant One of TextVariant to be used
```

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (sainthkh) [reply packages/block-serialization-default-parser/src/index.js]: The explanation `added optimization to prevent backtracking on attribute parsing` after `@since` tag isn't valid. 

Message below:

```
Invalid JSDoc @since: "4.6.1 added optimization to prevent backtracking on attribute parsing".eslint(jsdoc/check-values)
```

**Comment by sainthkh** [comment]
I used Record because adding `/* eslint-disable */` at the top and the bottom of JSDoc comment looks ugly and we don't know when this will be fixed. 

After reading your comment and related code, I agree that restoring `setting` is better than removing it. `setting` is important here.

But I don't think `key` in `{[key: string]: any}` doesn't provide much information. In this case, I think `Record` is better. For the same reason, I prefer `Record<WPKeycodeModifier, (key:string)=>any>` to `{[M in WPKeycodeModifier]:(key:string)=>any}`.

These are my opinion. If you think `Record` hurts consistency, I don't mind reverting them.

File: packages/components/src/text/mixins.js (hunk 2 above)
Comment (sainthkh) [reply packages/components/src/text/mixins.js]: Fixed.

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (aduth) [reply packages/block-serialization-default-parser/src/index.js]: Do you have any thoughts on how we might make this valid, while still maintaining the goals of including the message? I'm not sure if JSDoc has any standard way to describe updates in specific versions. It could be that we simply include this in the summary text.

e.g.

```
 * In 4.6.1, this was updated to add an optimization to prevent backtracking on
 * attribute parsing.
```

(Then the `@since 4.6.1` line could be removed)

Comment (sainthkh) [reply packages/block-serialization-default-parser/src/index.js]: How about this way?

```js
/** 
 *
 * @since 3.8.0
 * 
 * # History
 * - 4.6.1 added optimization to prevent backtracking on attribute parsing
 */
```

**Comment by sainthkh** [comment]
Rebased.

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (aduth) [reply packages/block-serialization-default-parser/src/index.js]: In lieu of any official JSDoc way of doing this, your suggestion seems reasonable to me.

A couple nit-picking points:

- As far as placement, it should probably be attached as part of the extended description. It might even be necessary to do so, to ensure that it's picked up by the parsing tools.
- We should test to see if the Markdown syntax works with how [docgen](https://github.com/WordPress/gutenberg/tree/master/packages/docgen) generates documentation, and if it does, whether we want it to be H1, since it's likely not the best heading level in the context of document. Depending if we want to _assume_ how the documentation is structured, it could either be H3 (`###`) or non-structured (bolded `**`).

>If we decide to use my proposal, then should we update the coding standard, too, for this?

It would be worth considering I think, yeah. And as such, it would be a good topic of a [future JavaScript meeting](https://make.wordpress.org/meetings/). I can see about adding it to the agenda for next week's meeting.

I was curious how this is managed in the WordPress PHP code. It appears the standard there is to add description as part of `@since`, just like what we're proposing to replace:

https://github.com/WordPress/wordpress-develop/blob/a44943546ca6259279f2632b1b4890bd7cc5ca22/src/wp-comments-post.php#L48-L49

Looking at the PHPDoc standard documentation for `@since`, I see this:

https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.since.pkg.html

What stands out to me is "[unspecified format]". It seems like something where the lack of a format imposed gives some potential flexibility by the project to decide to implement their own conventions.

But to come full circle, couldn't we do that as well? Looking at the JSDoc standard:

https://jsdoc.app/tags-since.html

There's not really an imposed restriction on what `<versionDescription>` could mean. You might even take "description" to mean that we _can_ include a summary of the changes. Maybe `eslint-plugin-jsdoc` is too strict, or it's something where we just want to opt out of this specific validation, because we're okay to include the summaries in the `@since` tag.

The more I think about it, the more I wonder if the simplest way to go about this is to just opt out of [`check-values`](https://github.com/gajus/eslint-plugin-jsdoc#check-values). What do you think?

**Comment by aduth** [comment]
> But I don't think `key` in `{[key: string]: any}` doesn't provide much information. In this case, I think `Record` is better. For the same reason, I prefer `Record<WPKeycodeModifier, (key:string)=>any>` to `{[M in WPKeycodeModifier]:(key:string)=>any}`.

I think it would depend. In this very specific case, you're right that `key` might not provide much value, because it's more or less a default term used when referring to the _key of an object_. But I don't think it would always be used this way.

Consider this example:

https://github.com/aduth/tannin/blob/cd1c7447843df7751c4abd1b92aee03fe56bfba4/packages/tannin/index.js#L38-L44

In that case, `domain` is a useful piece of information that would be lost by using `Record`.

That said, since we are talking about this specific case, I think it would be fine to use `Record`. We might want to consider whether it's a well-understood type (and as such, if we expect there to be any issue in developer onboarding).

This section of the current documentation is quite relevant, and might need to be updated accordingly:

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#record-types

**Comment by aduth** [comment]
I'm trying to understand what the specific difference is between `Object<string,string>` and `Record<string,string>`.

Related: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt

I have experienced in the past that TypeScript has a tendency to treat `Object` types as equivalent to the (dreaded) [`any` type](https://www.typescriptlang.org/docs/handbook/basic-types.html#any), which we most certainly would want to avoid. I guess `Record` is preferred in these cases where we want to loosely define the types of keys and values in what is expected to be provided as a plain object? I'm honestly not clear on this subject. If it is, that sounds quite nice actually.

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (sainthkh) [reply packages/block-serialization-default-parser/src/index.js]: I was confused with `check-value` and `check-type`. I thought we need both of them to check types. But after reading `check-value` doc, `check-value` is for `@version`, `@since`, `@author`, `@license` tags. 

We don't use `@author`, `@license`, `@version` tags. So, what `check-value` option only does for Gutenberg project is to ensure if the version number follows the semver format. We can check this manually. I think it doesn't harm to opt out of `check-values`.

The best scenario might be `check-value` has an option like 'allowVersionChanges` and checks the version format automatically. But it doesn't exist now. 

As for this, we need to change these 2 things [in the doc](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/):

**From:**

> @since x.x.x: Should always be 3-digit (e.g. @since 3.6.0).

**To:**

```
@since x.x.x: Should always be 3-digit (e.g. @since 3.6.0). You can optionally add what is changed after the number (e.g. @since 11.0.0 Check server state by default.)
```

To let new developers know that they can leave changes as comments if it's important. 

And add one more `@since`:

```js
/**
 * @since      x.x.x
 * @since      x.x.x Check server state by default
 */
```

Because some people don't read explanation above and go straight to the example.

**Comment by sainthkh** [comment]
In TypeScript, [`Object<string, number>`] doesn't exist. [Check this playground](http://www.typescriptlang.org/play/?ssl=5&ssc=11&pln=4&pc=1#code/DYUwLgBAhgXBDyAjAViAxmAPAZzAJwEsA7AcwBoIiBXAW0RDwD4IBeCAbwF8AobqAbQDkJAPYiAJoIC6rCABYATLygA6RFHGyFAdiA). (If it doesn't show any code, type in `let a: Object<string, number> = {}`, you'll know it soon.)

And according to [JSDoc doc](https://jsdoc.app/tags-type.html), the correct way to specifying types is `{Object.<string, number>}`. Requires `.` in the middle.

One difference between `Record` and indexable interface is that you can use `symbol` type as a key in `Record`. 

```ts
interface X {
    [x: symbol]: string // error
}

let a: Record<symbol, number> = {} // ok
```

But I guess it doesn't matter much. Because `symbol` isn't used a lot.

**Comment by aduth** [comment]
@sainthkh Would you suggest we update the documentation I linked in my previous comment to always recommend to use `Record` where otherwise we're using `Object`, then?

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (aduth) [reply packages/block-serialization-default-parser/src/index.js]: > As for this, we need to change these 2 things [in the doc](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/):
> 
> **From:**
> 
> > @SInCE x.x.x: Should always be 3-digit (e.g. @SInCE 3.6.0).
> 
> **To:**
> 
> ```
> @since x.x.x: Should always be 3-digit (e.g. @since 3.6.0). You can optionally add what is changed after the number (e.g. @since 11.0.0 Check server state by default.)
> ```
> 
> To let new developers know that they can leave changes as comments if it's important.
> 
> And add one more `@since`:
> 
> ```js
> /**
>  * @since      x.x.x
>  * @since      x.x.x Check server state by default
>  */
> ```
> 
> Because some people don't read explanation above and go straight to the example.

To me, that sounds like a very sensible change to the standard 👍 

I'll plan to add it to the agenda for next week's JavaScript meeting.

In the meantime, would it be a good course of action here then to opt out of `check-values`?

I think in a perfect world, we could validate our `@version` standard (either `x.x.x` or `x.x.x` plus a specially formatted message). I imagine would likely require building something custom, so not something I think needs to be done here.

Comment (aduth) [reply packages/block-serialization-default-parser/src/index.js]: I note there's a section about this tag in the PHP documentation standards:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#since-section-changelogs

**Comment by sainthkh** [comment]
My proposal:

* If the object key name is meaningful like `setting` or `domain`, use indexable interface.
* If not, use `Record`.

**Comment by aduth** [comment]
That seems sensible to me. Would you want to include those documentation changes here as well?

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (sainthkh) [reply packages/block-serialization-default-parser/src/index.js]: * I'll opt it out tomorrow. 
* I agree. Validating version/since should be a separate issue.

**Comment by sainthkh** [comment]
Yes. I think that's the part of this update.

**Comment by sainthkh** [comment]
@aduth 

* Opt out `jsdoc/check-value`. 
* Restored `setting` in `WPLinkControlSettingsValue`.
* Added `Record` rule to the coding guideline.
* Wrote #20859 for @since.

**Question:**

* According to [the guideline](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#record-types), it is incorrect to use `{Object}`, `{Function}`, `{Promise}`. But there is no rule to enforce it. Should we create one for this? If so, I'll create an issue for this, because there is none.
* If we decide to make a rule for them, it might be sensible to add an issue about removing them. In Gutenberg, there are 1304 `{Object}`s and 164 `{Function}`s and 34 `{Promise}`s.

**Comment by aduth** [comment]
> **Question:**
> 
> * According to [the guideline](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#record-types), it is incorrect to use `{Object}`, `{Function}`, `{Promise}`. But there is no rule to enforce it. Should we create one for this? If so, I'll create an issue for this, because there is none.
> * If we decide to make a rule for them, it might be sensible to add an issue about removing them. In Gutenberg, there are 1304 `{Object}`s and 164 `{Function}`s and 34 `{Promise}`s.

These are good questions, especially considering that the document is pretty unambiguous in calling them "Incorrect" (to clarify, I'm the one who wrote it this way). At least speaking from my own experience, I've not always been successful in following these rules so strictly, and I don't know that we want to put many barriers in place which might dissuade people from using types. At the same time, we should want to encourage people as much as possible to use the best options available to them. Some of these are easier to adopt than others. `Promise` is pretty straight forward to provide with a generic type, and we should encourage it always. `Function` is sometimes a much more accessible option than the technically-better-but-harder-to-write expanded form. In some cases, such as in a `@template` tag like with higher-order functions, it might even be fair to say it can be the _correct_ choice ([example](https://github.com/aduth/memize/blob/37a2422b79c92f0113b36be5c165f589c2d869b9/index.js#L33-L38))?

So to specifically answer your question:

- Right now, I don't think we should want to enforce this.
- The documentation should perhaps be updated to _encourage_ moreso than codify (phrasing like "Bad" vs. "Good" or "Not Ideal" vs. "Better", etc).
- We may still want to audit existing code to see where we can make updates. I'd be a bit hesitant that an issue would be the most effective way to track this, compared to simply setting a good precedent and clear documentation for people to make good judgment in their own refactoring.

I think any or all of these could change over time as we continue to explore how to implement types in the codebase. I'm also open to arguments to the contrary.

File: packages/block-serialization-default-parser/src/index.js (hunk 1 above)
Comment (aduth) [reply packages/block-serialization-default-parser/src/index.js]: To follow up, this topic was discussed during today's JavaScript meeting in the development Slack:

https://make.wordpress.org/core/2020/03/17/javascript-chat-summary-march-17-2020/

It was well-received. I don't anticipate there being any problems to update the documentation standards, and I'll plan to do so later in the week, assuming there is no opposition raised in the comments of the summary above.

Aside: You're welcomed to join our weekly chats if you'd like! They're held each Tuesday in the Slack instance ([meeting calendar](https://make.wordpress.org/meetings/), [chat registration](https://make.wordpress.org/chat/)).

**Review by aduth — COMMENTED** [review]
This is looking good!

One last thing I see is that the Coding Guidelines document still has quite a few references to `Object` types throughout. Should these be updated to `Record`, at least when used as a generic?

Related to my previous comment, there are also some interesting cases where it's not so clear what the best type would be. Notably, for React function components, how best should we document the `props` argument.

For example:

```js
/**
 * Renders a toolbar.
 *
 * @param {Object} props             Component props.
 * @param {string} [props.className] Class to set on the container `<div />`.
 */
```

Is `{Object}` okay here? Is it something where we want to try to use the "indexable interface" format for describing the keys of the `props` object (I don't know that this is feasible since, unlike TypeScript, we're confined to describing the type in a single line)? I had wondered if the `props` `@param` line even provides any value, vs. just omitting and including only the specific properties. Last time I tried that, I think there were some errors related to the missing `props` documentation.

**Comment by sainthkh** [comment]
1\. 

>In some cases, such as in a @template tag like with higher-order functions, it might even be fair to say it can be the correct choice (example)?

If we're using TypeScript, the type of `F` might be `(...args: Array<any>) => any`. It's a longer version of `Function`. We don't need to or can't use that in Gutenberg.

2\. 

> Is {Object} okay here?

It seems that we have no other option for now.

```js
/**
 *
 * @param {number} x xx
 * @param {number} y yy
 */
export const x = ( { x, y } ) => {
	return x + y;
};
```

When we write a simple function like above, TypeScript type checker will show us this error message:

```
packages/a11y/src/addContainer.js:44:20 - error TS8024: JSDoc '@param' tag has name 'y', but there is no parameter with that name.

44  * @param {number} y yy
                      ~

packages/a11y/src/addContainer.js:46:22 - error TS2339: Property 'x' does not exist on type 'Number'.

46 export const x = ( { x, y } ) => {
                        ~

packages/a11y/src/addContainer.js:46:25 - error TS2339: Property 'y' does not exist on type 'Number'.

46 export const x = ( { x, y } ) => {
```

3\.

> Is it something where we want to try to use the "indexable interface" format for describing the keys of the props object (I don't know that this is feasible since, unlike TypeScript, we're confined to describing the type in a single line)?

I think there's not much difference. And we all know that it's a React function by its name starting with a capital case. This is not a dictionary-like object that we want to add/remove elements on the run time. In addition, we don't need more `/* eslint-disable jsdoc/check-type */`.

4\.

I checked the guideline and `{Object}` was used 4 times.

* Custom type definition -> It's OK there. It's not a dictionary.
* Generic `{Object}` or object as a dictionary -> It's changed to `Record`.
* Optional item in props object -> It's OK. Because it's not a dictionary.
* React component typing -> Same as above. 

5\.

The best to way to enforce types is to migrate to TS and use `strict` option. But it's a big project and it's outside the topic of this project. 

But like eslint warnings, when types aren't enforced, people ignore them. Maybe we need to discuss it later. (I know you had a meeting about TypeScript. And I also know that it can't be a priority now.)

**Comment by aduth** [comment]
>When we write a simple function like above, TypeScript type checker will show us this error message:

In your example, I sense the individual properties would still need to include some object name in order to use the dot notation:

https://jsdoc.app/tags-param.html#parameters-with-properties

I was hoping it would work with just an arbitrary name (props, for a function component):

```js
/**
 * Returns a person's name.
 *
 * @type {import('react').FC}
 *
 * @param {string} props.name Person name.
 */
function Person( { name } ) {
	return <div>{ name }</div>;
}
```

But the ESLint rule doesn't like it:

```
@param path declaration ("props.name") appears before any real parameter.eslint(jsdoc/check-param-names)
```

---

>Custom type definition -> It's OK there. It's not a dictionary.

Are you referring to [this one](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#custom-types)?

```js
/**
 * A block selection object.
 *
 * @typedef {Object} WPBlockSelection
 *
 * @property {string} clientId     A block client ID.
 * @property {string} attributeKey A block attribute key.
 * @property {number} offset       An attribute value offset, based on the rich
 *                                 text value.
 */
```

If I recall correctly, we can remove `{Object}` altogether there, and it's understood to mean a type with those other properties. In fact, I feel I've run into issues where _keeping_ `{Object}` can be problematic in how that type is used (I can't recall specifically what they were).

---

>Optional item in props object -> It's OK. Because it's not a dictionary.

Which is this referring to?

---

>It seems that nothing should be changed.

What about [this one](https://github.com/WordPress/gutenberg/blob/37be2e9cd5b336a87d40cea46b4e92e94bafa637/docs/contributors/coding-guidelines.md#documenting-examples)

```
 * @return {Object<string,WPDataSelector>} Object containing the store's
 *                                         selectors.
```

Should this be `{Record<string,WPDataSelector>}`, or no?

---

>But like eslint warnings, when types aren't enforced, people ignore them. Maybe we need to discuss it later. (I know you had a meeting about TypeScript. And I also know that it can't be a priority now.)

There's definitely a desire to enforce types inferred from JSDoc. See #18838. It's already in place for a handful of modules. It's not quite to the same extent as going full TypeScript, but it is enforced where modules have been opted-in, and I'm highly in favor of more modules to be enabled.

---

**To summarize**, the only points above that I think need some decision or action:

- Can we remove `{Object}` from ` * @typedef {Object} WPBlockSelection` in the documentation?
- Can we convert `* @return {Object<string,WPDataSelector>}` to `{Record<string,WPDataSelector>}` in the documentation?

**Comment by sainthkh** [comment]
I searched `{Object}` and missed `{Object<`. 

> Can we remove {Object} from * @typedef {Object} WPBlockSelection in the documentation?

Yes. eslint and tsc don't complain about this. And `@property` lines below succinctly says that it's a type for an object (or `interface` in TypeScript jargon).

> Can we convert * @return {Object<string,WPDataSelector>} to {Record<string,WPDataSelector>} in the documentation?

Yes. `Object<T, K>` doesn't exist. And it's a dictionary of items with special type. So, `Record` is right here. 

>> Optional item in props object -> It's OK. Because it's not a dictionary.
>
> Which is this referring to?

This one below:

```js
/**
 * Renders a toolbar.
 *
 * @param {Object} props             Component props.
 * @param {string} [props.className] Class to set on the container `<div />`.
 */
```

I experimented if we can remove `{Object}` here like `@typedef`. But TypeScript wants it. 

-----

It seems that everything is OK. I'll go fix the coding guideline.

**Review by aduth — APPROVED** [review]
Thanks for the work here, and for your patience and well-reasoned thoughts during the many iterations. This looks good to land now 🎉
